Moonshot/baddies/BaddieLaser/BaddieSpawner.gd

```diff
@@ -0,0 +1,53 @@
+extends Position2D
+class_name BaddieSpawner
+
+var bearBoi = load("res://baddies/characters/BearBoi.tscn")
+var baddie_builder = load("res://procedural_map_generation/BaddieBuilder.gd")
+var max_baddies = 5 setget set_max_baddies
+var _baddie_dictionary: Dictionary = {}
+
+func set_max_baddies(m: int) -> void:
+	max_baddies = m
+
+func spawn_randomly() -> void:
+	spawn()
+	var timer: Timer = Timer.new()
+	add_child(timer)
+	timer.start(0.5)
+
+func spawn():
+	var baddie_dictionary_keys: Array = _baddie_dictionary.keys()
+	if _is_max_capacity(baddie_dictionary_keys):
+		return
+
+	var builder = baddie_builder.new(bearBoi, { "inital_hp": 0.75, "hp": 0.75 })
+	var baddie_instance = builder.build()
+
+	baddie_instance.position = global_position -  Vector2(0.75, 0.75) # slightly beneath mother hen
+	baddie_instance.scale = Vector2(0.75, 0.75) # smaller
+
+	_baddie_dictionary[baddie_instance.get_instance_id()] = baddie_instance
+
+	_add_baddie_to_scene(baddie_instance)
+	_clean_dictionary(baddie_dictionary_keys)
+
+func _is_max_capacity(baddie_dictionary_keys: Array) -> bool:
+	return baddie_dictionary_keys.size() > max_baddies
```

Comment:
Maybe when spawned baddies are killed they're removed from the dictionary so more can spawn. Not sure if this is an easy change however.

Reply:
`_clean_dictionary ` does this:

![image](https://user-images.githubusercontent.com/29375815/100107525-a501dc00-2e61-11eb-8774-7c34e98464ea.png)


 This is an implementation of a [ring buffer](https://en.wikipedia.org/wiki/Circular_buffer) to ensure the total number of bear boi's never exceeds `max_baddies `. 

And when a free space is available it gets filled